### DISCUSSION TITLE
Difference between dev.langchain4j.service.DefaultAiServices<T> & io.quarkiverse.langchain4j.QuarkusAiServicesFactory.QuarkusAiServices<T>

### DISCUSSION THREAD

**Discussion by gkfirst8** [quarkiverse/quarkus-langchain4j#702] [Q&A]
Hi,

I was started working with Quarkus-langchain4j again after contributing to langchain4j in January. I contributed the "fromResource" to the SystemMessage & UserMessage. These are handled by the DefaultAiServices.

How should I go about using these features when using Quarkus-langchain4j? Quarkus-langchain4j defaults to using the QuarkusAiServices, which makes sense because it probably needs to do some special stuff in the context of Quarkus.

But it looks like I'm missing all the 'goodies' developed in DefaultAiServices. Is there a way to get those back?

with kind regads,
Gerke

**Comment by geoand** [comment]
> Quarkus-langchain4j defaults to using the QuarkusAiServices, which makes sense because it probably needs to do some special stuff in the context of Quarkus.

Very true! We need our own implementation because we want to support build time processing.

> But it looks like I'm missing all the 'goodies' developed in DefaultAiServices. Is there a way to get those back?

What goodies do you mean? In any case, it means introducing the support in `QuarkusAiServices`.

**Reply by gkfirst8** [reply]
Hi @geoand , thanks for your reply.

I was specifically referring to the added support of fromResource to the System- & User-Message. I can feel that it might more work work in the QuarkusAiServices as it needs build-time support?

The fromResource allows the prompt to be stored in a resource rather than directly in the code. This would make creating and passing around the prompt between different experts easier.

Any idea if we can still re-use some of the code without the need to fully re-implement this? Maybe we can extend the original DefaultAiService somehow? (Currently the SystemMessage & UserMessage are re-used but are missing the support for the extra parameter(s) (does QuarkusAiSerices support the delimiter ?).)

**Reply by gkfirst8** [reply]
(Come to think of it, maybe I should have moved support for fromResource from DefaultAiServices to AiServices ?)

**Reply by geoand** [reply]
> I was specifically referring to the added support of fromResource to the System- & User-Message. I can feel that it might more work work in the QuarkusAiServices as it needs build-time support?

Those should already work in Quarkus LangChain4j. If they don't, I'm interested in  sample that shows they don't.
